tools/swift-format/Sources/Rules/GroupNumericLiterals.swift

```diff
@@ -10,7 +10,67 @@ import SwiftSyntax
 ///
 /// Format: All numeric literals that should be grouped will have `_`s inserted where appropriate.
 ///
+/// TODO: Minimum numeric literal length bounds and numeric groupings selected arbitrarily, could
+///       be  reevaluated.
+///
+/// TODO: Handle floating point literals
+///
 /// - SeeAlso: https://google.github.io/swift#numeric-literals
 public final class GroupNumericLiterals: SyntaxFormatRule {
+  public override func visit(_ node: IntegerLiteralExprSyntax) -> ExprSyntax {
+    
+    var strDigits = node.digits.text
+    guard !strDigits.contains("_") else { return node }
+    
+    let isNegative = strDigits.first == "-"
+    strDigits = isNegative ? String(strDigits.dropFirst()) : strDigits
+    
+    var newDigits = ""
+    
+    switch strDigits.prefix(2) {
+    case "0x":
+      // Hexadecimal
+      let digitsNoPrefix = String(strDigits.dropFirst(2))
+      guard let intDigits = Int(digitsNoPrefix, radix: 16) else { return node }
+      guard intDigits >= 0x1000_0000 else { return node }
+      diagnose(.groupNumericLiteral(by: 4), on: node)
+      newDigits = "0x" + groupDigits(digitStr: digitsNoPrefix, by: 4)
+    case "0b":
+      // Binary
+      let digitsNoPrefix = String(strDigits.dropFirst(2))
+      guard let intDigits = Int(digitsNoPrefix, radix: 2) else { return node }
+      guard intDigits >= 0b1_000000000 else { return node }
+      diagnose(.groupNumericLiteral(by: 8), on: node)
+      newDigits = "0b" + groupDigits(digitStr: digitsNoPrefix, by: 8)
+    case "0o":
+      // Octal
+      return node
+    default:
+      // Decimal
+      guard let intDigits = Int(strDigits) else { return node }
+      guard intDigits >= 1_000_000 else { return node }
+      diagnose(.groupNumericLiteral(by: 3), on: node)
+      newDigits = groupDigits(digitStr: strDigits, by: 3)
+    }
+    
+    newDigits = isNegative ? "-" + newDigits : newDigits
+    return node.withDigits(SyntaxFactory.makeIdentifier(newDigits))
+  }
+  
+  func groupDigits(digitStr: String, by: Int) -> String {
```

Comment:
Some naming recommendations:

* Avoid abbreviations like "Str" in names; `digits` by itself would be fine.
* Parameter names should be nouns/noun phrases (except for Booleans, which are verb phrases), so if you have a function that uses a preposition as a label, like `by`, give the argument variable an explicit name. For example, `by stride: Int`. Then use `stride` inside the function body instead of `by`.